### PR TITLE
refactor: move component helper to the $compileProvider

### PR DESCRIPTION
**Pull request by Narretz**
The Module.component() now delegates to $compileProvider.component().
This has the following benefits:
- when using only the loader, we are not accessing out of scope variables / functions
- components can be registered via $compileProvider
- docs are a bit easier to find

/cc @petebacondarwin @shahata

### CODE REVIEW

File: test/ng/compileSpec.js [hunk 1]
```diff
@@ -9276,4 +9276,159 @@ describe('$compile', function() {
...
+      module('my');
+
+      inject(function($compile, $rootScope, log) {
+        element = $compile('<my-component></my-component>')($rootScope);
+        expect(element.find('div').text()).toEqual('SUCCESS');
+        expect(log).toEqual('OK');
+      });
+    });
+
+    it('should register a directive via $compileProvider.component()', function() {
```
Comment (Narretz) [inline test/ng/compileSpec.js]: new test

File: test/ng/compileSpec.js [hunk 2]
```diff
@@ -9276,4 +9276,159 @@ describe('$compile', function() {
...
     });
   });
+
+  describe('component helper', function() {
+    it('should return the module', function() {
+      var myModule = angular.module('my', []);
+      expect(myModule.component('myComponent', {})).toBe(myModule);
+    });
+
+    it('should register a directive', function() {
```
Comment (Narretz) [inline test/ng/compileSpec.js]: this test is new

File: src/ng/compile.js [hunk 3]
```diff
@@ -915,6 +915,162 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
...
+   *   });
+   * ```
+   *
+   * <br />
+   * See also {@link ng.$compileProvider#directive $compileProvider.directive()}.
+   */
+  this.component = function registerComponent(name, options) {
+    function factory($injector) {
+      function makeInjectable(fn) {
+        if (isFunction(fn) || Array.isArray(fn)) {
```
Comment (gkalpak) [inline src/ng/compile.js]: Better to use `isArray()` now that we're here.

File: src/ng/compile.js [hunk 4]
```diff
@@ -915,6 +915,162 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
...
+        }
+      }
+
+      var template = (!options.template && !options.templateUrl ? '' : options.template);
+      return {
+        controller: options.controller || function() {},
+        controllerAs: identifierForController(options.controller) || options.controllerAs || name,
+        template: makeInjectable(template),
+        templateUrl: makeInjectable(options.templateUrl),
+        transclude: options.transclude === undefined ? false : options.transclude,
```
Comment (gkalpak) [inline src/ng/compile.js]: Minor, but `options.transclude` or `options.transclude || false` would work as well (now that `false` is the default).

**Comment by gkalpak** [comment]
A couple of minor nitpicks, but LGTM otherwise.

<sub>This will (obviously) affect the fix for https://github.com/angular/batarang/issues/275.</sub>

**Comment by petebacondarwin** [comment]
LGTM - merging with tweaks suggested by @gkalpak and some minor docs changes

**Pull request closed without merging**